mockwebserver/src/main/java/okhttp3/mockwebserver/MockWebServer.kt

```diff
@@ -0,0 +1,1114 @@
+      sink.writeUtf8("\r\n")
+      i++
+    }
+    sink.writeUtf8("\r\n")
+    sink.flush()
+  }
+
+  private fun sleepIfDelayed(delayMs: Long) {
+    if (delayMs != 0L) {
+      try {
+        Thread.sleep(delayMs)
+      } catch (e: InterruptedException) {
+        throw AssertionError(e)
+      }
+    }
+  }
+
+  /**
+   * Transfer bytes from [source] to [sink] until either [byteCount] bytes have
+   * been transferred or [source] is exhausted. The transfer is throttled according to [policy].
+   */
+  @Throws(IOException::class)
+  private fun throttledTransfer(
+    policy: MockResponse,
+    socket: Socket,
+    source: BufferedSource,
+    sink: BufferedSink,
+    byteCount: Long,
+    isRequest: Boolean
+  ) {
+    var byteCountNum = byteCount
+    if (byteCountNum == 0L) return
+
+    val buffer = Buffer()
+    val bytesPerPeriod = policy.throttleBytesPerPeriod
+    val periodDelayMs = policy.getThrottlePeriod(TimeUnit.MILLISECONDS)
+
+    val halfByteCount = byteCountNum / 2
+    val disconnectHalfway = if (isRequest)
+      policy.getSocketPolicy() === DISCONNECT_DURING_REQUEST_BODY
+    else
+      policy.getSocketPolicy() === DISCONNECT_DURING_RESPONSE_BODY
+
+    while (!socket.isClosed) {
+      var b = 0L
+      while (b < bytesPerPeriod) {
+        // Ensure we do not read past the allotted bytes in this period.
+        var toRead = min(byteCountNum, bytesPerPeriod - b)
+        // Ensure we do not read past halfway if the policy will kill the connection.
+        if (disconnectHalfway) {
+          toRead = min(toRead, byteCountNum - halfByteCount)
+        }
+
+        val read = source.read(buffer, toRead)
+        if (read == -1L) return
+
+        sink.write(buffer, read)
+        sink.flush()
+        b += read
+        byteCountNum -= read
+
+        if (disconnectHalfway && byteCountNum == halfByteCount) {
+          socket.close()
+          return
+        }
+
+        if (byteCountNum == 0L) return
+      }
+
+      if (periodDelayMs != 0L) {
+        try {
+          Thread.sleep(periodDelayMs)
+        } catch (e: InterruptedException) {
+          throw AssertionError(e)
+        }
+      }
+    }
+  }
+
+  @Throws(IOException::class)
+  private fun readEmptyLine(source: BufferedSource) {
+    val line = source.readUtf8LineStrict()
+    if (line.isNotEmpty()) throw IllegalStateException("Expected empty but was: $line")
+  }
+
+  override fun toString(): String = "MockWebServer[$port]"
+
+  @Throws(IOException::class)
+  override fun close() = shutdown()
+
+  /** A buffer wrapper that drops data after [bodyLimit] bytes.  */
+  private class TruncatingBuffer internal constructor(
+    private var remainingByteCount: Long
+  ) : Sink {
+    internal val buffer = Buffer()
+    internal var receivedByteCount = 0L
+
+    @Throws(IOException::class)
+    override fun write(source: Buffer, byteCount: Long) {
+      val toRead = min(remainingByteCount, byteCount)
+      if (toRead > 0) {
```

Comment:
mind changing this to 0L? I’m trying to get us more formal in all of these

Reply:
Fixed.